File: server/src/utils/sessionStore.js
// sessionStore.js
//
// Simple in-memory store, intentionally not a database — this is a demo/take-home project.
// Swapping this for Redis or Postgres would mean implementing the same five functions.

const sessions = new Map();

function ensure(sessionId) {
  if (!sessions.has(sessionId)) {
    sessions.set(sessionId, {
      id: sessionId,
      createdAt: new Date().toISOString(),
      messages: [], // chat transcript: {role, content}
      reasoningLog: [], // every log() entry, for admin replay
      decisions: [], // finalized outcomes: approve/deny/escalate
      customerContext: null, // last-known verified customer, for the admin sidebar
    });
  }
  return sessions.get(sessionId);
}

export function createSession(sessionId) {
  return ensure(sessionId);
}

export function getSession(sessionId) {
  return sessions.get(sessionId) || null;
}

export function listSessions() {
  return Array.from(sessions.values())
    .map((s) => ({
      id: s.id,
      createdAt: s.createdAt,
      messageCount: s.messages.length,
      lastDecision: s.decisions[s.decisions.length - 1] || null,
      customerContext: s.customerContext,
    }))
    .sort((a, b) => new Date(b.createdAt) - new Date(a.createdAt));
}

export function appendMessage(sessionId, message) {
  const s = ensure(sessionId);
  s.messages.push(message);
  return s;
}

export function appendLog(sessionId, entry) {
  const s = ensure(sessionId);
  s.reasoningLog.push(entry);
  return s;
}

export function appendDecision(sessionId, decision) {
  const s = ensure(sessionId);
  s.decisions.push(decision);
  return s;
}

export function setCustomerContext(sessionId, customer) {
  const s = ensure(sessionId);
  s.customerContext = customer
    ? {
        customer_id: customer.customer_id,
        name: customer.name,
        email: customer.email,
        tier: customer.tier,
        fraud_flag: customer.fraud_flag,
      }
    : null;
  return s;
}

export function deleteSession(sessionId) {
  sessions.delete(sessionId);
}

File: fastapi_server/utils/session_store.py
from datetime import datetime

sessions = {}

def ensure(session_id: str):
    if session_id not in sessions:
        sessions[session_id] = {
            "id": session_id,
            "createdAt": datetime.utcnow().isoformat() + "Z",
            "messages": [],
            "reasoningLog": [],
            "decisions": [],
            "customerContext": None
        }
    return sessions[session_id]

def create_session(session_id: str):
    return ensure(session_id)

def get_session(session_id: str):
    return sessions.get(session_id)

def list_sessions():
    s_list = []
    for s in sessions.values():
        s_list.append({
            "id": s["id"],
            "createdAt": s["createdAt"],
            "messageCount": len(s["messages"]),
            "lastDecision": s["decisions"][-1] if s["decisions"] else None,
            "customerContext": s["customerContext"]
        })
    return sorted(s_list, key=lambda x: x["createdAt"], reverse=True)

def append_message(session_id: str, message: dict):
    s = ensure(session_id)
    s["messages"].append(message)
    return s

def append_log(session_id: str, entry: dict):
    s = ensure(session_id)
    s["reasoningLog"].append(entry)
    return s

def append_decision(session_id: str, decision: dict):
    s = ensure(session_id)
    s["decisions"].append(decision)
    return s

def set_customer_context(session_id: str, customer: dict):
    s = ensure(session_id)
    if customer:
        s["customerContext"] = {
            "customer_id": customer.get("customer_id"),
            "name": customer.get("name"),
            "email": customer.get("email"),
            "tier": customer.get("tier"),
            "fraud_flag": customer.get("fraud_flag")
        }
    else:
        s["customerContext"] = None
    return s

def delete_session(session_id: str):
    if session_id in sessions:
        del sessions[session_id]
